examples/tic-tac-toe/cli/src/client.rs

```diff
@@ -0,0 +1,749 @@
+        Ok(Self {
+            wallet,
+            package: conn.package_id,
+        })
+    }
+
+    /// Fetch the details of a game object from on-chain (can be either shared or owned).
+    pub(crate) async fn game(&self, id: ObjectID) -> Result<Game> {
+        let client = self.client().await?;
+
+        // (1) Read from RPC
+        let response = client
+            .read_api()
+            .get_object_with_options(
+                id,
+                SuiObjectDataOptions {
+                    show_owner: true,
+                    show_bcs: true,
+                    ..Default::default()
+                },
+            )
+            .await
+            .context("Error fetching game over RPC")?;
+
+        if let Some(err) = response.error {
+            bail!(err);
+        }
+
+        // (2) Perform validation checks
+        let Some(SuiObjectData {
+            object_id,
+            version,
+            digest,
+            bcs: Some(raw),
+            owner: Some(owner),
+            ..
+        }) = response.data
+        else {
+            bail!("No data for game");
+        };
+
+        let Some(raw) = raw.try_as_move() else {
+            bail!("It is a package, not a object.");
+        };
+
+        if raw.type_.name.as_str() != "Game" {
+            bail!("It is not a Game object, it has type {}", raw.type_);
+        }
+
+        let package = ObjectID::from(raw.type_.address);
+        if package != self.package {
+            bail!(
+                "It is expected to be from package {} but has type {}",
+                self.package,
+                raw.type_,
+            );
+        }
+
+        // (3) Deserialize contents
+        let kind = match raw.type_.module.as_str() {
+            "shared" => GameKind::Shared(
+                bcs::from_bytes(&raw.bcs_bytes).context("Failed to deserialize contents")?,
+            ),
+
+            "owned" => GameKind::Owned(
+                bcs::from_bytes(&raw.bcs_bytes).context("Failed to deserialize contents")?,
+            ),
+
+            kind => bail!("{id} has unrecognised Game kind: {kind}"),
+        };
+
+        // (4) Check whether the game has ended or not.
+        let mut builder = ProgrammableTransactionBuilder::new();
+        let g = if let Owner::Shared {
+            initial_shared_version,
+        } = owner
+        {
+            builder.obj(ObjectArg::SharedObject {
+                id,
+                initial_shared_version,
+                mutable: false,
+            })?
+        } else {
+            builder.obj(ObjectArg::ImmOrOwnedObject((object_id, version, digest)))?
+        };
+
+        builder.programmable_move_call(
+            self.package,
+            raw.type_.module.clone(),
+            Identifier::new("ended").unwrap(),
+            vec![],
+            vec![g],
+        );
+
+        let results = client
+            .read_api()
+            .dev_inspect_transaction_block(
```

Comment:
Is this `dev_inspect_transaction_block` used before actually executing the last move to determine if this last move leads to the end of game?

Reply:
This dev inspect call is used to determine if the game has already ended. We run it every time we load the Game. When we make a move, we issue the transaction and then we read the Game state, so at that point we will see whether the game has ended, to display that.